addons/l10n_ro_edi_etransport/models/l10n_ro_edi_etransport_document.py

```diff
@@ -0,0 +1,563 @@
+
+    @api.depends('l10n_ro_edi_etransport_operation_type_code')
+    def _compute_end_loc_type_index(self):
+        for doc in self:
+            doc.l10n_ro_edi_etransport_end_loc_type_index = doc._get_location_type_field_index('end')
+
+    def _check_values(self):
+        self.ensure_one()
+        errors = []
+
+        # operation type
+        if not self.l10n_ro_edi_etransport_operation_type_id:
+            errors.append(_("Operation type is missing."))
+            return errors  # return prematurely because a lot of fields depend on the operation type
+
+        # operation scope
+        if not self.l10n_ro_edi_etransport_operation_scope_id:
+            errors.append(_("Operation scope is missing."))
+
+        # vehicle & trailer numbers
+        if not self.l10n_ro_edi_etransport_vehicle_number:
+            errors.append(_("Vehicle number is missing."))
+
+        # All filled-in vehicle and trailer numbers must be unique
+        license_plates = [num for num in (self.l10n_ro_edi_etransport_vehicle_number, self.l10n_ro_edi_etransport_trailer_1_number, self.l10n_ro_edi_etransport_trailer_2_number) if num]
+        if len(license_plates) != len(set(license_plates)):
+            errors.append(_("Vehicle number and trailer number fields must be unique."))
+
+        # rate codes
+        if self.l10n_ro_edi_etransport_operation_type_code not in ('60', '70'):
+            product_without_code_names = {move_line.product_id.name for move in self.l10n_ro_edi_etransport_move_ids
+                                          for move_line in move.move_line_ids
+                                          if not move_line.product_id.intrastat_code_id.code}
+
+            if product_without_code_names:
+                if len(product_without_code_names) == 1:
+                    (product_name,) = product_without_code_names
+                    errors.append(_("Product %(name)s is missing the intrastat code value.", name=product_name))
+                else:
+                    errors.append(_("Products %(names)s are missing the intrastat code value.", names=", ".join(product_without_code_names)))
+
+        # Location types
+        def check_location_type(location: str) -> str | None:
+            loc_type_idx = getattr(self, f'l10n_ro_edi_etransport_{location}_loc_type_index')
+            if loc_type_idx not in (1, 2):
+                return
+
+            if not getattr(self, f'l10n_ro_edi_etransport_{location}_loc_type_{loc_type_idx}'):
+                match location:
+                    case 'start':
+                        return _("'Start Location Type' is missing")
+                    case 'end':
+                        return _("'End Location Type' is missing")
+
+        for loc in ('start', 'end'):
+            if loc_error := check_location_type(loc):
+                errors.append(loc_error)
+                return errors  # return prematurely because all the location fields depend on these fields
```

Comment:
inner function is a bit ugly IMO, especially when they're only called for one/two times. Why not just check immediately inside the for loop?